pkg/common-controller/snapshot_controller.go

```diff
@@ -1374,3 +1391,61 @@ func (ctrl *csiSnapshotCommonController) setAnnVolumeSnapshotBeingDeleted(conten
 	}
 	return nil
 }
+
+// addInvalidContentLabel adds a label to invalid content objects.
+func (ctrl *csiSnapshotCommonController) checkAndAddInvalidContentLabel(content *crdv1.VolumeSnapshotContent) error {
```

Comment:
this function could cause any API update calls after invoking it to error out if the content is invalid and updated in API server.
specifically, a contentClone will be created and used to update with the API server, however the original "content" has not been
updated. later update calls using this "content" will likely fail because the version is different now.
maybe update the passed in "content" reference to the updated one in cache?

Reply:
In general, using Patch instead of Update helps reduce failed updates due to conflicting versions, if different components are updating different fields.

Reply:
Changing the update calls to patch feels a bit out of scope for this (already big) PR. I changed the function to return the updated ptr if it was changed, otherwise the passed in one.

This doesn't work https://github.com/kubernetes-csi/external-snapshotter/blob/c436f5ccb3c1942890caf8c6b839678613d577e7/pkg/common-controller/snapshot_controller.go#L1366 because it's changing the value of the local variable. See https://play.golang.org/p/Qu70fnDEP7N. Should I fix it?

Reply:
> @AndiLi99 please fix existing invalid tests which are failing due to newly added invalid labels.

There was a tricky version bug caused by not using the return value of Update() combined with tests which didn't specify either of spec.source fields which was causing my unit tests to fail strangely.